File: Android/AnyMind/app/src/main/java/com/anymind/anymind/util/TagParser.kt
package com.anymind.anymind.util

object TagParser {
    private val tagRegex = Regex("#([\\p{L}\\p{N}_-]+)")
    private val systemTagSet = setOf("temp", "longterm", "p1", "p2", "conflict")

    fun extractTags(content: String): List<String> {
        val tags = tagRegex.findAll(content).mapNotNull { match ->
            val raw = match.groupValues.getOrNull(1)?.lowercase()?.trim()
            if (raw.isNullOrEmpty()) null else "#" + raw
        }.toSet()
        return tags.sorted()
    }

    fun splitTags(tags: List<String>): Pair<List<String>, List<String>> {
        val system = mutableSetOf<String>()
        val user = mutableSetOf<String>()
        for (tag in tags) {
            val trimmed = tag.trim().lowercase()
            val name = if (trimmed.startsWith("#")) trimmed.substring(1) else trimmed
            if (systemTagSet.contains(name)) {
                system.add("#" + name)
            } else if (name.isNotEmpty()) {
                user.add("#" + name)
            }
        }
        return system.sorted() to user.sorted()
    }

    fun isSystemTag(tag: String): Boolean {
        val trimmed = tag.trim().lowercase()
        val name = if (trimmed.startsWith("#")) trimmed.substring(1) else trimmed
        return systemTagSet.contains(name)
    }

    fun isTagOnlyLine(line: String): Boolean {
        val trimmed = line.trim()
        if (trimmed.isEmpty()) {
            return true
        }
        val stripped = tagRegex.replace(trimmed, "")
        val cleaned = stripped.replace(",", " ").trim()
        return cleaned.isEmpty()
    }

    fun parseFilterInput(input: String): List<String> {
        val tags = extractTags(input)
        if (tags.isNotEmpty()) {
            return tags
        }
        val tokens = input.split(',', ' ', '\n', '\t')
            .map { it.trim().lowercase() }
            .filter { it.isNotEmpty() }
        return tokens.map { if (it.startsWith("#")) it else "#" + it }.toSet().sorted()
    }
}

File: ios/AnyMind/PromptRecorder/Helpers/TagParser.swift
import Foundation

enum TagParser {
    private static let tagPattern = "#([\\p{L}\\p{N}_-]+)"
    private static let systemTagSet: Set<String> = ["temp", "longterm", "p1", "p2", "conflict"]

    static func extractTags(from content: String) -> [String] {
        guard let regex = try? NSRegularExpression(pattern: tagPattern, options: []) else {
            return []
        }
        let range = NSRange(content.startIndex..<content.endIndex, in: content)
        let matches = regex.matches(in: content, options: [], range: range)
        let tags = matches.compactMap { match -> String? in
            guard let tagRange = Range(match.range(at: 1), in: content) else { return nil }
            let raw = String(content[tagRange]).lowercased()
            guard !raw.isEmpty else { return nil }
            return "#" + raw
        }
        return Array(Set(tags)).sorted()
    }

    static func splitTags(_ tags: [String]) -> (system: [String], user: [String]) {
        var system: [String] = []
        var user: [String] = []
        for tag in tags {
            let trimmed = tag.trimmingCharacters(in: .whitespacesAndNewlines)
            let normalized = trimmed.lowercased()
            let name = normalized.hasPrefix("#") ? String(normalized.dropFirst()) : normalized
            if systemTagSet.contains(name) {
                system.append("#" + name)
            } else {
                user.append("#" + name)
            }
        }
        return (Array(Set(system)).sorted(), Array(Set(user)).sorted())
    }

    static func isSystemTag(_ tag: String) -> Bool {
        let trimmed = tag.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        let name = trimmed.hasPrefix("#") ? String(trimmed.dropFirst()) : trimmed
        return systemTagSet.contains(name)
    }

    static func isTagOnlyLine(_ line: String) -> Bool {
        let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty {
            return true
        }
        guard let regex = try? NSRegularExpression(pattern: tagPattern, options: []) else {
            return false
        }
        let range = NSRange(trimmed.startIndex..<trimmed.endIndex, in: trimmed)
        let stripped = regex.stringByReplacingMatches(in: trimmed, options: [], range: range, withTemplate: "")
        let cleaned = stripped
            .replacingOccurrences(of: ",", with: " ")
            .replacingOccurrences(of: "，", with: " ")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        return cleaned.isEmpty
    }

    static func parseFilterInput(_ input: String) -> [String] {
        let tags = extractTags(from: input)
        if !tags.isEmpty {
            return tags
        }
        let tokens = input
            .split { $0 == "," || $0.isWhitespace }
            .map { String($0).lowercased() }
            .filter { !$0.isEmpty }
        return Array(Set(tokens.map { $0.hasPrefix("#") ? $0 : "#" + $0 })).sorted()
    }
}
